source/validate_image.cpp

```diff
@@ -0,0 +1,1356 @@
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Expected operand to be of type OpTypeImage: "
+            << spvOpcodeString(opcode);
+      }
+      break;
+    }
+
+    case SpvOpImageQuerySizeLod: {
+      if (!_.IsIntScalarOrVectorType(result_type)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Expected Result Type to be int scalar or vector type: "
+            << spvOpcodeString(opcode);
+      }
+
+      const uint32_t image_type = _.GetOperandTypeId(inst, 2);
+      if (_.GetIdOpcode(image_type) != SpvOpTypeImage) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Expected Image to be of type OpTypeImage: "
+            << spvOpcodeString(opcode);
+      }
+
+      ImageTypeInfo info;
+      if (!GetImageTypeInfo(_, image_type, &info)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Corrupt image type definition";
+      }
+
+      uint32_t expected_num_components = info.arrayed;
+      switch (info.dim) {
+        case SpvDim1D:
+          expected_num_components += 1;
+          break;
+        case SpvDim2D:
+        case SpvDimCube:
+          expected_num_components += 2;
+          break;
+        case SpvDim3D:
+          expected_num_components += 3;
+          break;
+        default:
+          return _.diag(SPV_ERROR_INVALID_DATA)
+              << "Image 'Dim' must be 1D, 2D, 3D or Cube: "
+              << spvOpcodeString(opcode);
+      };
+
+      if (info.multisampled != 0) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Image 'MS' must be 0: " << spvOpcodeString(opcode);
+      }
+
+      uint32_t result_num_components = _.GetDimension(result_type);
+      if (result_num_components != expected_num_components) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Result Type has " << result_num_components << " components, "
+            << "but " << expected_num_components << " expected: "
+            << spvOpcodeString(opcode);
+      }
+
+      const uint32_t lod_type = _.GetOperandTypeId(inst, 3);
+      if (!_.IsIntScalarType(lod_type) && !_.IsFloatScalarType(lod_type)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Expected Level of Detail to be int or float scalar: "
+            << spvOpcodeString(opcode);
+      }
+
+      break;
+    }
+
+    case SpvOpImageQuerySize: {
+      if (!_.IsIntScalarOrVectorType(result_type)) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Expected Result Type to be int scalar or vector type: "
+            << spvOpcodeString(opcode);
+      }
+
+      const uint32_t image_type = _.GetOperandTypeId(inst, 2);
+      if (_.GetIdOpcode(image_type) != SpvOpTypeImage) {
+        return _.diag(SPV_ERROR_INVALID_DATA)
+            << "Expected Image to be of type OpTypeImage: "
+            << spvOpcodeString(opcode);
+      }
+
+#if 0
+      // TODO(atgoo@github.com) The spec doesn't whitelist all Dims supported by
```

Comment:
??
SPIR-V 1.2 seems to whitelist the dimensionality in the description of the Image argument:

>  Its Dim operand must be one of Rect or Buffer, or if its MS is 1, it can be 2D, or, if its Sampled Type is 0 or 2, it can be 2D or 3D. 

So it's Rect, Buffer, 2D, 3D.

The same list is implied by the description of the result type.

Reply:
But it doesn't whitelist all of them:
https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/imageSize.xhtml